content/technotes/2020-02-26-tic-ghactions.Rmd

```diff
@@ -0,0 +1,170 @@
+In tic we use [r-lib/setup-r](https://github.com/r-lib/actions/tree/master/setup-r) as the base combined with [r-lib/setup-pandoc](https://github.com/r-lib/actions/tree/master/setup-pandoc) and [r-lib/setup-tinytex](https://github.com/r-lib/actions/tree/master/setup-tinytex).
+We apply some opinionated changes on top of which we think enhance the CI setup substantially.
+In the following we explain these changes to be as transparent as possible.
+
+#### Caching
+
+By default [r-lib/actions](https://github.com/r-lib/actions) caches all direct dependencies of the package which is being tested on a successful run.
+This is done via action [actions/cache](https://github.com/actions/cache) by hashing a list of all package dependencies:
+
+```{yml}
+- name: Query dependencies
+  run: |
+    install.packages('remotes')
+    saveRDS(remotes::dev_package_deps(dependencies = TRUE), "depends.Rds", version = 2)
+  shell: Rscript {0}
+
+- name: Cache R packages
+  if: runner.os != 'Windows'
+  uses: actions/cache@v1
+  with:
+    path: ${{ env.R_LIBS_USER }}
+    key: ${{ runner.os }}-r-${{ matrix.config.r }}-${{ hashFiles('depends.Rds') }}
+    restore-keys: ${{ runner.os }}-r-${{ matrix.config.r }}-
+```
+
+tic instead caches the complete R library of the run **on every run** by using a forked version of [actions/cache](https://github.com/actions/cache): [pat-s/always-upload-cache](https://github.com/pat-s/always-upload-cache).
+
+These differences have the advantage that packages which are needed for side-effects (like building a pkgdown site or running codecov) are cached too.
+Because caches are immutable, tic (needs to) rebuilds the cache **daily**.
+If this would not happen, every runner would otherwise always use the first cache forever.
+When packages are updated on CRAN, tic would download the cache first but then update those before proceeding to the "script" stage.
+After some time all packages from the cache would be outdated, making the cache useless.
+Hence, tic rebuilds the cache daily.
+Users can change this behavior to their liking by changing the bash code that is used to generate the timestamps used for caching:
+
+```{yml}
+- name: "[Cache] Prepare daily timestamp for cache"
+  if: runner.os != 'Windows'
+  id: date
+  run: echo "::set-output name=date::$(date '+%d-%m')"
+
+- name: "[Cache] Prepare weekly timestamp for cache"
+  if: runner.os != 'Windows'
+  id: datew
+  run: echo "::set-output name=datew::$(date '+%Y-%V')"
+```
+
+The cache rebuilding happens in a scheduled CRON run daily at 3 am in the morning.
+Because tic always uploads the cache even if the build failed ([r-lib/actions](https://github.com/r-lib/actions) only uploads on a successful build), there will always be a cache available for builds which start during the day.
+So even if the build of a repository takes several minutes to install all dependencies, the first commit being made every day will always make use of an existing cache.
```

Comment:
I have to temporarily for now because "scheduled" events are not supported atm due to an internal change from GitHub. Will be available in the future though. I'll try to make that clear.

Reply:
See my comment above